<a href="https://colab.research.google.com/github/rwiederblank/Ruchama_Wiederblank_MCON232_module1/blob/main/MCON232_FQ_Homework_1_Lists_Strings_FileIO.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# MCON 232 FQ • Homework 1 • Lists (1D) + Strings + File I/O (Colab-friendly)

**Estimated time:** ~75 minutes  
**Due:** _(your date here)_  
**Name:** ___________________________

### Rules (important)
- **No list comprehensions** for this homework.
- **No slicing** (like `lst[1:4]`) unless a problem explicitly tells you to.
- Use **loops** and **basic list methods** (`append`, `insert`, `pop`, `remove`).
- For string cleanup, use: `strip()` and `split()`.
- Keep your code readable: meaningful variable names, blank lines, and short comments.

> Tip: Run the notebook **top to bottom**. Some later parts rely on earlier cells.


In [1]:
# Optional: quick check that you're running in a notebook environment
print("Notebook is running ✅")


Notebook is running ✅


## Part 1 — Refresher: Lists & Strings (10 minutes)

These are quick warmups to get your hands moving:
- `len(list)`
- indexing (first / last)
- updating list elements
- basic list methods
- string cleanup with `strip()` and `split()`


### Task 1A — Fix the list (easy warmup)

You will practice:
- length
- first / last element
- replacing an element


In [2]:
animals = ["cat", "dog", "fish", "bird"]

# TODO:
# 1) Print the length of the list
# 2) Print the first item
# 3) Print the last item (no slicing)
# 4) Replace "fish" with "hamster"
# 5) Print the updated list

# Write your code below:
print(len(animals))
print(animals[0])
print(animals[-1])
animals[2] = "hamster"
print(animals)


4
cat
bird
['cat', 'dog', 'hamster', 'bird']


### Task 1B — List methods practice

You **must** use the method named in each step.


In [4]:
nums = [10, 20, 30]

# TODO:
# 1) append 40
# 2) insert 99 at index 1
# 3) pop the last value into a variable called last_val
# 4) remove 20
# 5) print nums and last_val

# Write your code below:
nums.append(40)
nums.insert(1,99)
last_val = nums.pop()
nums.remove(20)
print(nums)
print(last_val)

[10, 99, 30]
40


### Task 1C — String refresher: `strip()` and `split()`

A file line often looks like: `"Alice,90\n"`  
We want to clean it and break it into parts.

Example:
- `line.strip()` removes whitespace/newlines at both ends
- `clean.split(",")` splits into a list of strings


In [5]:
line = "Alice,90\n"

# TODO:
# 1) Create a variable clean_line that removes the newline using strip()
# 2) Split clean_line by comma into a list called parts
# 3) Print clean_line and parts

# Write your code below:
clean_line = line.strip()
parts = line.split(",")
print(clean_line)
print(parts)

Alice,90
['Alice', '90\n']


## Part 2 — Thinking with Loops (20 minutes)

These problems are about **patterns**:
- counting
- filtering into a new list
- summing and multiplying with accumulators
- tracking an index with a loop


### Task 2A — Count & filter (no comprehensions)

Given a list of numbers:
1) Count how many are **>= 10**
2) Build a new list containing only numbers **>= 10**


In [6]:
nums = [3, 8, 2, 11, 5, 10, 10, 1]

# Required:
# count_big  -> how many numbers >= 10
# big_nums   -> list of numbers >= 10

count_big = 0
big_nums = []
for num in nums:
  if num >= 10:
    big_nums.append(num)

# TODO: write your loop here

print("count_big:", count_big)
print("big_nums:", big_nums)


count_big: 0
big_nums: [11, 10, 10]


### Task 2B — Two accumulators

Compute:
1) `total_sum` of the list
2) `even_product` = product of ONLY the even numbers  
   - Start `even_product` at 1  
   - Multiply only when the number is even


In [7]:
nums = [4, 7, 2, 9, 10, 3, 8]

total_sum = 0
even_product = 1

# TODO: write your loop here
for num in nums:
  total_sum += num
  if num%2==0:
    even_product*=num

print("total_sum:", total_sum)
print("even_product:", even_product)


total_sum: 43
even_product: 640


### Task 2C — “Where is it?” (index tracking)

Find the index of the **first** `"hello"` using a loop.  
Do **not** use `.index()` for this one.


In [9]:
words = ["hi", "hello", "hey", "hello", "hola"]

first_hello_index = -1  # keep -1 if not found

# TODO: write your loop here
# Hint: use range(len(words)) so you can check indexes.
for i in range(len(words)):
  if words[i] == "hello":
    first_hello_index = i
    break

print("first_hello_index:", first_hello_index)


first_hello_index: 1


## Part 3 — File I/O Refresher (Colab-Friendly) (25 minutes)

In Google Colab, your notebook runs on a temporary computer.  
So we will **write a file first**, then **read it back**.

We will work with a file named: `grades.txt`

Each line contains:
```
Name,Score
```
Example:
```
Alice,90
```


### Step 3A — Create the input file (provided)

Run this cell **as-is** (do not change it).


In [10]:
# DO NOT CHANGE THIS CELL
sample_text = """Alice,90
Ben,72
Cara,88
Dan,95
Eva,72
"""

with open("grades.txt", "w") as f:
    f.write(sample_text)

print("grades.txt created ✅")


grades.txt created ✅


### Step 3B — Read and parse the file into lists

Build two **parallel lists**:
- `names`  (strings)
- `scores` (ints)

Requirements:
- Use a loop to read each line
- Use `strip()` to remove the newline
- Use `split(",")` to separate name and score
- Convert score to `int`


In [12]:
names = []
scores = []

# TODO:
# 1) Open grades.txt for reading
# 2) Loop through each line
# 3) Strip newline
# 4) Split by comma into name and score
# 5) Append name to names
# 6) Convert score to int and append to scores

# Write your code below:
with open("grades.txt", "r") as f:
  for line in f:
    line = line.strip()
    line_split = line.split(",")
    names.append(line_split[0])
    scores.append(line_split[1])

print("names:", names)
print("scores:", scores)


names: ['Alice', 'Ben', 'Cara', 'Dan', 'Eva']
scores: ['90', '72', '88', '95', '72']


### Step 3C — Analyze the data (thinking)

Using `names` and `scores`, compute:

1) `avg_score` (average score as a float)
2) `highest_score` and the `highest_name` who earned it
3) `count_72` = how many students earned exactly 72
4) `passing_names` = list of students with score **>= 75**

**Do not use list comprehensions.**


In [19]:
# Required variables:
avg_score = 0.0
total = 0
count = 0
highest_score = -1
highest_name = ""
count_72 = 0
passing_names = []

# TODO:
# - compute total
# - compute avg_score
# - find highest score + name
# - count 72s
# - build passing_names
for i in range(len(names)):
  total+= int(scores[i])
  count += 1
  if i == 0 or scores[i] > highest_score:
    highest_score = scores[i]
    highest_name = names[i]
  if int(scores[i]) == 72:
    count_72 += 1
  if int(scores[i]) >= 75:
    passing_names.append(names[i])

avg_score = total / count

print("avg_score:", avg_score)
print("highest:", highest_name, highest_score)
print("count_72:", count_72)
print("passing_names:", passing_names)


avg_score: 83.4
highest: Dan 95
count_72: 2
passing_names: ['Alice', 'Cara', 'Dan']


### Step 3D — Write a report file

Create a file named `report.txt` that contains lines like:

- `Average: 83.4`
- `Top Student: Dan (95)`
- `Count of 72: 2`
- `Passing: Alice, Cara, Dan, Eva`

Then read and print the contents of `report.txt` to prove it worked.


In [26]:
# TODO:
# 1) Open report.txt in write mode and write the 4 lines
# 2) Then open report.txt in read mode and print it
with open("report.txt", "w") as f:
  f.write(f"Avergae: {avg_score} ")
  f.write(f"Top Student: {highest_name} ({highest_score}) ")
  f.write(f"Count of 72: {count_72} ")
  f.write(f"Passing: {passing_names} ")

with open("report.txt", "r") as f:
  line = f.readline()
  print(line)
# Write your code below:


Avergae: 83.4 Top Student: Dan (95) Count of 72: 2 Passing: ['Alice', 'Cara', 'Dan'] 


## Part 4 — “Thinky” Problem: Remove Duplicates (20 minutes)

You are not using sets yet, so we’ll solve this with a loop.

**Goal:** Build `unique_nums` containing each number only once, in the order it first appears.

Example:
- Input:  `[2, 5, 2, 9, 5, 1]`
- Output: `[2, 5, 9, 1]`

Hint:  
Use this pattern:
- for each `n` in `nums`  
  - if `n` is **not** already in `unique_nums`, append it


In [27]:
nums = [2, 5, 2, 9, 5, 1, 9, 9, 3]

unique_nums = []

for num in nums:
  if num in unique_nums:
    continue
  else:
    unique_nums.append(num)

print("unique_nums:", unique_nums)


unique_nums: [2, 5, 9, 1, 3]


### Optional (bonus) — How many duplicates were removed?

Compute `duplicates_removed`.

Hint:
- original count is `len(nums)`
- unique count is `len(unique_nums)`


In [28]:
# BONUS:
duplicates_removed = 0

# TODO:
duplicates_removed = len(nums) - len(unique_nums)

print("duplicates_removed:", duplicates_removed)


duplicates_removed: 4


## Submit

When finished, **download your notebook** and submit it according to your class instructions.

✅ You should have completed:
- basic list operations
- loop patterns (count, filter, accumulate, track index)
- string cleanup with `strip()` and parsing with `split()`
- file I/O (write → read → analyze → write report)
- a simple algorithmic thinking problem (remove duplicates)
